go/vt/vtgate/plan_execute.go

```diff
@@ -57,58 +78,96 @@ func (e *Executor) newExecute(
+			}
+		}
 
-	if err != nil {
-		safeSession.ClearWarnings()
-		return err
-	}
+		vcursor, err := newVCursorImpl(safeSession, comments, e, logStats, e.vm, vs, e.resolver.resolver, e.serv, e.warnShardedOnly, e.pv)
+		if err != nil {
+			return err
+		}
 
-	if plan.Type != sqlparser.StmtShow {
-		safeSession.ClearWarnings()
-	}
+		// 3: Create a plan for the query
+		var plan *engine.Plan
+		// If we are retrying, it is likely that the routing rules have changed and hence we need to replan the query since the target
+		// keyspace of the resolved shards may have changed.
+		plan, err = e.getPlan(ctx, vcursor, query, stmt, comments, bindVars, reservedVars, e.normalize, logStats)
+		execStart := e.logPlanningFinished(logStats, plan)
 
-	// add any warnings that the planner wants to add
-	for _, warning := range plan.Warnings {
-		safeSession.RecordWarning(warning)
-	}
+		if err != nil {
+			safeSession.ClearWarnings()
+			return err
+		}
 
-	result, err := e.handleTransactions(ctx, mysqlCtx, safeSession, plan, logStats, vcursor, stmt)
-	if err != nil {
-		return err
-	}
-	if result != nil {
-		return recResult(plan.Type, result)
-	}
+		if plan.Type != sqlparser.StmtShow {
+			safeSession.ClearWarnings()
+		}
 
-	// 3: Prepare for execution
-	err = e.addNeededBindVars(vcursor, plan.BindVarNeeds, bindVars, safeSession)
-	if err != nil {
-		logStats.Error = err
-		return err
-	}
+		// add any warnings that the planner wants to add
+		for _, warning := range plan.Warnings {
+			safeSession.RecordWarning(warning)
+		}
 
-	if plan.Instructions.NeedsTransaction() {
-		return e.insideTransaction(ctx, safeSession, logStats,
-			func() error {
-				return execPlan(ctx, plan, vcursor, bindVars, execStart)
-			})
-	}
+		result, err := e.handleTransactions(ctx, mysqlCtx, safeSession, plan, logStats, vcursor, stmt)
+		if err != nil {
+			return err
+		}
+		if result != nil {
+			return recResult(plan.Type, result)
+		}
 
-	return execPlan(ctx, plan, vcursor, bindVars, execStart)
+		// 4: Prepare for execution
+		err = e.addNeededBindVars(vcursor, plan.BindVarNeeds, bindVars, safeSession)
+		if err != nil {
+			logStats.Error = err
+			return err
+		}
+
+		// 5: Execute the plan and retry if needed
+		if plan.Instructions.NeedsTransaction() {
+			err = e.insideTransaction(ctx, safeSession, logStats,
+				func() error {
+					return execPlan(ctx, plan, vcursor, bindVars, execStart)
+				})
+		} else {
+			err = execPlan(ctx, plan, vcursor, bindVars, execStart)
+		}
+		if err != nil {
+			rootCause := vterrors.RootCause(err)
+			if rootCause != nil && strings.Contains(rootCause.Error(), "enforce denied tables") {
+				log.Infof("%d:%t will retry query %s due to %v", try, retrying, query, err)
+				retrying = true
+				lastVSchemaCreated = vs.GetCreated()
+				continue
+			} else {
+				retrying = false
+			}
+		} else {
+			retrying = false
+		}
+		if err == nil && try > 0 {
+			log.Infof("query %d:%t succeeded on retry: %s", try, retrying, query)
+		}
```

Comment:
when a query is executed inside a transaction, then we cannot retry, instead we should rollback the transaction as it cannot be completed.

Reply:
Good catch, this has been fixed.